### DISCUSSION TITLE
Using Types of Members of Generic Containers

### DISCUSSION THREAD

**Discussion by hmc-cs-mdrissi** [python/typing#1087] [Q&A] [answered]
I currently have code that uses a generic config like class that is processed by various other classes. Sometimes those other classes there types depend on what types contained by config class. Something like,

```python
class Config(Generic[T1, T2]):
  foo: dict[str, T1]
  bar: list[T1]
  baz: T2
  # ... (more shared non-generic elements)

ConfigT = TypeVar('ConfigT', bound=Config)

class Processor(Generic[ConfigT]):
  def __init__(self, config: ConfigT):
    self._config = config

  def process_foo(self) -> list[XYZ]:
    return list(self._config.foo.values())
```

The main issue is filling in `XYZ`. One option I tried was making processor generic on T1, T2 too, but https://github.com/python/typing/issues/548 I think is a blocker to write ConfigT[T1, T2]. Are there any alternative ways to have classes operate on top of generic classes like this? For now I'm just falling back to XYZ is Any. The processor class mainly cares for the inner types of foo/bar/baz/etc because of relationships they hold. Since same typevar T1 appears multiple times in Config class so even though it may know little about T1 it would ideally type check things like 
```python
x = self.process_foo()
x.extend(self._config.bar)
``` 
is good but 
```python
x = self.process_foo()
x.append(self._config.baz)
``` 
is an error.

**Comment by erictraut** [comment, marked as answer]
How about this?

```python
class Processor(Generic[T1, T2]):
    def __init__(self, config: Config[T1, T2]):
        self._config = config

    def process_foo(self) -> list[T1]:
        return list(self._config.foo.values())
```

**Reply by hmc-cs-mdrissi** [reply]
That looks like a good trade off and works better than making ConfigT generic. The one issue it has is I also have subclasses of Processor that may only work for certain configs (something like),

```python
class ConfigX(Config):
  ...

class ProcessorX(Processor[ConfigX]):
  ...
```

but I can handle that by having subclasses override the type of `config: ...` A liskov error, but I think liskov error and proper checking of T1/T2 is better than T1/T2 being uncheckable. It also pushes me hard to try to minimize logic specific to config subclasses.
